From 931e6b0a2b1d0dbe1b8f6b420d80c4b77b5d0f3b Mon Sep 17 00:00:00 2001
From: Jayachandran B <jayachandran.b@intel.com>
Date: Wed, 25 Feb 2015 10:33:51 +0530
Subject: [PATCH 05/15] ASoC: Intel: Enabling Audio Offload and Deep buffer PB
 in CHT

Issue: GMINL-6292

This Patch enables audio offload and deepbuffer for CHT.

Change-Id: I5786f38675d8a2fa17403575457201e6a84dd059
Signed-off-by: Jayachandran B <jayachandran.b@intel.com>
Signed-off-by: Praveen Diwakar <praveen.diwakar@intel.com>
Signed-off-by: Divya Prakash <divya1.prakash@intel.com>
Signed-off-by: Fabien Marotte <fabien.marotte@intel.com>
Signed-off-by: Samreen Nilofer <samreen.nilofer@intel.com>
---
 arch/x86/include/asm/platform_cht_audio.h        |  2 ++
 sound/core/compress_offload.c                    |  3 +++
 sound/soc/intel/board/cht_bl_dpcm_rt5672.c       | 31 ++++++++++++++++++++++++
 sound/soc/intel/pcm.c                            | 27 +++++++++++++++++++++
 sound/soc/intel/platform-libs/controls_v2_dpcm.c |  2 +-
 sound/soc/intel/sst/sst_platform.c               |  4 +++
 6 files changed, 68 insertions(+), 1 deletion(-)

diff --git a/arch/x86/include/asm/platform_cht_audio.h b/arch/x86/include/asm/platform_cht_audio.h
index 51965d7..8a29405 100644
--- a/arch/x86/include/asm/platform_cht_audio.h
+++ b/arch/x86/include/asm/platform_cht_audio.h
@@ -14,6 +14,8 @@
 
 enum {
 	CHT_DPCM_AUDIO = 0,
+	CHT_DPCM_DB,
+	CHT_DPCM_COMPR,
 	CHT_DPCM_VOIP,
 };
 
diff --git a/sound/core/compress_offload.c b/sound/core/compress_offload.c
index 7233a30..e41bbfd 100644
--- a/sound/core/compress_offload.c
+++ b/sound/core/compress_offload.c
@@ -856,6 +856,9 @@ static const struct file_operations snd_compr_file_ops = {
 		.write =	snd_compr_write,
 		.read =		snd_compr_read,
 		.unlocked_ioctl = snd_compr_ioctl,
+#ifdef CONFIG_COMPAT
+		.compat_ioctl =	snd_compr_ioctl,
+#endif
 		.mmap =		snd_compr_mmap,
 		.poll =		snd_compr_poll,
 };
diff --git a/sound/soc/intel/board/cht_bl_dpcm_rt5672.c b/sound/soc/intel/board/cht_bl_dpcm_rt5672.c
index 4e4c255..181e330 100644
--- a/sound/soc/intel/board/cht_bl_dpcm_rt5672.c
+++ b/sound/soc/intel/board/cht_bl_dpcm_rt5672.c
@@ -554,6 +554,11 @@ static int cht_aif1_hw_params(struct snd_pcm_substream *substream,
 	return 0;
 }
 
+static int cht_compr_set_params(struct snd_compr_stream *cstream)
+{
+	return 0;
+}
+
 static const struct snd_soc_pcm_stream cht_dai_params = {
 	.formats = SNDRV_PCM_FMTBIT_S24_LE,
 	.rate_min = 48000,
@@ -747,6 +752,9 @@ static struct snd_soc_ops cht_be_ssp2_ops = {
 	.hw_params = cht_aif1_hw_params,
 };
 
+static struct snd_soc_compr_ops cht_compr_ops = {
+	.set_params = cht_compr_set_params,
+};
 
 static struct snd_soc_dai_link cht_dailink[] = {
 	[CHT_DPCM_AUDIO] = {
@@ -761,6 +769,29 @@ static struct snd_soc_dai_link cht_dailink[] = {
 		.dynamic = 1,
 		.ops = &cht_aif1_ops,
 	},
+	[CHT_DPCM_DB] = {
+		.name = "Cherrytrail DB Audio Port",
+		.stream_name = "Deep Buffer Audio",
+		.cpu_dai_name = "Deepbuffer-cpu-dai",
+		.codec_dai_name = "snd-soc-dummy-dai",
+		.codec_name = "snd-soc-dummy",
+		.platform_name = "sst-platform",
+		.ignore_suspend = 1,
+		.dynamic = 1,
+		.ops = &cht_aif1_ops,
+		.dpcm_playback = 1,
+	},
+	[CHT_DPCM_COMPR] = {
+		.name = "Cherrytrail Compressed Port",
+		.stream_name = "Cherrytrail Compress",
+		.cpu_dai_name = "Compress-cpu-dai",
+		.codec_dai_name = "snd-soc-dummy-dai",
+		.codec_name = "snd-soc-dummy",
+		.platform_name = "sst-platform",
+		.ignore_suspend = 1,
+		.dynamic = 1,
+		.compr_ops = &cht_compr_ops,
+	},
 	[CHT_DPCM_VOIP] = {
 		.name = "Cherrytrail VOIP Port",
 		.stream_name = "Cherrytrail Voip",
diff --git a/sound/soc/intel/pcm.c b/sound/soc/intel/pcm.c
index ade3f81..39c0c34 100644
--- a/sound/soc/intel/pcm.c
+++ b/sound/soc/intel/pcm.c
@@ -564,6 +564,10 @@ static struct snd_soc_dai_ops sst_media_dai_ops = {
 	.mute_stream = sst_media_digital_mute,
 };
 
+static struct snd_soc_dai_ops sst_compr_dai_ops = {
+	.mute_stream = sst_media_digital_mute,
+};
+
 static struct snd_soc_dai_ops sst_be_dai_ops = {
 	.startup = sst_enable_ssp,
 	.shutdown = sst_disable_ssp,
@@ -589,6 +593,17 @@ static struct snd_soc_dai_driver sst_platform_dai[] = {
 	},
 },
 {
+	.name = SST_DEEPBUFFER_DAI,
+	.ops = &sst_media_dai_ops,
+	.playback = {
+		.stream_name = "Deepbuffer Playback",
+		.channels_min = SST_STEREO,
+		.channels_max = SST_STEREO,
+		.rates = SNDRV_PCM_RATE_48000,
+		.formats = SNDRV_PCM_FMTBIT_S16_LE,
+	},
+},
+{
 	.name = SST_SPEAKER_DAI,
 	.ops = &sst_media_dai_ops,
 	.playback = {
@@ -600,6 +615,18 @@ static struct snd_soc_dai_driver sst_platform_dai[] = {
 	},
 },
 {
+	.name = SST_COMPRESS_DAI,
+	.compress_dai = 1,
+	.ops = &sst_compr_dai_ops,
+	.playback = {
+		.stream_name = "Compress Playback",
+		.channels_min = SST_STEREO,
+		.channels_max = SST_STEREO,
+		.rates = SNDRV_PCM_RATE_48000,
+		.formats = SNDRV_PCM_FMTBIT_S16_LE,
+	},
+},
+{
 	.name = SST_VOIP_DAI,
 	.ops = &sst_media_dai_ops,
 	.playback = {
diff --git a/sound/soc/intel/platform-libs/controls_v2_dpcm.c b/sound/soc/intel/platform-libs/controls_v2_dpcm.c
index e7fda03..73e18ab 100644
--- a/sound/soc/intel/platform-libs/controls_v2_dpcm.c
+++ b/sound/soc/intel/platform-libs/controls_v2_dpcm.c
@@ -1553,7 +1553,7 @@ static const struct snd_soc_dapm_widget sst_dapm_widgets[] = {
 	/* MediaX IN paths are set via ALLOC, so no SET_MEDIA_PATH command */
 	SST_PATH_INPUT("media0_in", SST_TASK_MMX, SST_SWM_IN_MEDIA0, sst_generic_modules_event),
 	SST_PATH_INPUT("media1_in", SST_TASK_MMX, SST_SWM_IN_MEDIA1, NULL),
-	SST_PATH_INPUT("media2_in", SST_TASK_MMX, SST_SWM_IN_MEDIA2, sst_set_media_path),
+	SST_PATH_INPUT("media2_in", SST_TASK_MMX, SST_SWM_IN_MEDIA2, NULL),
 	SST_PATH_INPUT("media3_in", SST_TASK_MMX, SST_SWM_IN_MEDIA3, NULL),
 	SST_PATH_OUTPUT("media0_out", SST_TASK_MMX, SST_SWM_OUT_MEDIA0, sst_set_media_path),
 	SST_PATH_OUTPUT("media1_out", SST_TASK_MMX, SST_SWM_OUT_MEDIA1, sst_set_media_path),
diff --git a/sound/soc/intel/sst/sst_platform.c b/sound/soc/intel/sst/sst_platform.c
index 9459cda..1339a45 100644
--- a/sound/soc/intel/sst/sst_platform.c
+++ b/sound/soc/intel/sst/sst_platform.c
@@ -40,6 +40,10 @@ static struct sst_dev_stream_map dpcm_strm_map_cht[] = {
 	{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}, /* Reserved, not in use */
 	{CHT_DPCM_AUDIO, 0, SNDRV_PCM_STREAM_PLAYBACK, PIPE_RSVD,
 					SST_TASK_ID_MEDIA, SST_DEV_MAP_IN_USE},
+	{CHT_DPCM_DB,    0, SNDRV_PCM_STREAM_PLAYBACK, PIPE_RSVD,
+					SST_TASK_ID_MEDIA, SST_DEV_MAP_IN_USE},
+	{CHT_DPCM_COMPR, 0, SNDRV_PCM_STREAM_PLAYBACK, PIPE_RSVD,
+					SST_TASK_ID_MEDIA, SST_DEV_MAP_IN_USE},
 	{CHT_DPCM_VOIP,  0, SNDRV_PCM_STREAM_PLAYBACK, PIPE_RSVD,
 					SST_TASK_ID_MEDIA, SST_DEV_MAP_IN_USE},
 	{CHT_DPCM_AUDIO, 0, SNDRV_PCM_STREAM_CAPTURE, PIPE_RSVD,
-- 
1.9.1

